### PR TITLE
[WIP] Tighten Lvalues and remove all references to Node

### PR DESCRIPTION
**Pull request by elazarg**
<s>TypeMap is tighten to Dict[Expression, Type], which will help removing most of the remaining references to Node. It will also allow moving `literal` and `literal_hash` from Node to Expression (leaving only line and column as data in Node), possibly making it easier to perform the NB regarding TypeMap (see inside).

One culprit is that TypeMap is Optional everywhere now. Admittedly I don't understand why is it Optional at all.</s>

[I intend to use this PR as a reference for changes in much smaller chunks]

All references to Node are removed from everywhere except mypy.node.

Make Lvalue a separate Union type. This has many implications; I think the PR can serve as a demonstration for now.

The need for dedicated LvalueTuple and LvalueStar is seen through the casts needed when accessing their elements. Perhaps making them generic can help.

### CODE REVIEW

**Comment by rwbarton** [comment]
Sorry the name is bad, but `TypeMap` isn't just supposed to be a generic map from expressions to types, but is specifically a thing returned by `find_isinstance_check` that represents the knowledge about types we gain from knowing a certain expression is true (or false). It can be `None` when that case is impossible (e.g., an unreachable branch). There's a big comment explaining this before its definition.

I haven't been following what you've been doing here closely, but I think it doesn't make sense to merge `TypeMap` with other places you see `Dict[Node, Type]`. It's probably better to just change both `TypeMap` and other `Dict[Node, Type]`s to `Dict[Expression, Type]` and change as little else as possible.

**Comment by elazarg** [comment]
Thanks @rwbarton. I have noticed (after posting this PR) that there are different maps with different purposes. I still think there should be a single name for each purpose, but I am not that certain.

File: mypy/binder.py [hunk 1]
```diff
@@ -1,22 +1,27 @@
...
+from mypy.nodes import (Expression, Var, RefExpr, SymbolTableNode)
 
 from mypy.subtypes import is_subtype
 from mypy.join import join_simple
 from mypy.sametypes import is_same_type
 
+from mypy.nodes import IndexExpr, MemberExpr, NameExpr
 
-class Frame(Dict[Any, Type]):
-    pass
```
Comment (elazarg) [inline mypy/binder.py]: I don't understand why is it needed. This can be fixed regardless of this PR.

File: mypy/binder.py [hunk 2]
```diff
@@ -165,21 +175,20 @@ def pop_frame(self, fall_through: int = 0) -> Frame:
 
         return result
 
-    def get_declaration(self, node: Node) -> Type:
-        if isinstance(node, (RefExpr, SymbolTableNode)) and isinstance(node.node, Var):
-            type = node.node.type
+    def get_declaration(self, expr: BLval) -> Type:
+        if isinstance(expr, (RefExpr, SymbolTableNode)) and isinstance(expr.node, Var):
+            type = expr.node.type
```
Comment (elazarg) [inline mypy/binder.py]: The test here is never true, and I think the original (changed in one of my previous PRs) was never true either. I wonder why.

File: mypy/checker.py [hunk 3]
```diff
@@ -1812,7 +1820,7 @@ def visit_decorator(self, e: Decorator) -> Type:
                 if d.fullname == 'typing.no_type_check':
                     e.var.type = AnyType()
                     e.var.is_ready = True
-                    return NoneTyp()
+                    return None
```
Comment (elazarg) [inline mypy/checker.py]: This mistake can be fixed regardless of this PR.

**Comment by gvanrossum** [comment]
Will you let us know when you're ready for a review?

**Comment by elazarg** [comment]
@gvanrossum this PR is ready now, as far as I'm concerned. However, since it is so big, and since introducing Lvalue as a separate type has implications on the old parser, I am splitting it to smaller chunks, without "real" Lvalue yet. I will close this one and will submit a new PR momentarily.

I also have certain questions about the specific details of the code, and about design decisions, but I don't know what's the right place to ask them; they are probably too minor for dedicated issues.

**Pull request closed without merging**